tokenizers/src/processors/template.rs

```diff
@@ -0,0 +1,704 @@
+///     .sequence_b(vec!["$1", "[SEP]"])
+///     .special_tokens(vec![("[CLS]", 1), ("[SEP]", 0)])
+///     .build()
+///     .unwrap();
+/// ```
+///
+#[derive(Debug, Clone, PartialEq, Builder, Serialize, Deserialize)]
+#[serde(tag = "type", from = "TemplateProcessingDeserializer")]
+#[builder(build_fn(validate = "Self::validate"))]
+pub struct TemplateProcessing {
+    #[builder(setter(into), default = "self.default_seq(0)")]
+    sequence_a: Template,
+    #[builder(setter(into), default = "self.default_seq(1)")]
+    sequence_b: Template,
+    #[builder(setter(skip), default = "self.default_added(true)")]
+    #[serde(skip)]
+    added_a: usize,
+    #[builder(setter(skip), default = "self.default_added(false)")]
+    #[serde(skip)]
+    added_b: usize,
+    #[builder(setter(into), default)]
+    special_tokens: Tokens,
+}
+
+/// We use this custom deserializer to provided the values for `added_a` and `added_b`
+/// during deserialization, while not having to serialize them
+#[doc(hidden)]
+#[derive(Deserialize)]
+#[serde(tag = "type")]
+struct TemplateProcessingDeserializer {
+    sequence_a: Template,
+    sequence_b: Template,
+    special_tokens: Tokens,
+}
+impl From<TemplateProcessingDeserializer> for TemplateProcessing {
+    fn from(t: TemplateProcessingDeserializer) -> Self {
+        let added_a = count_added(&t.sequence_a, Some(&t.special_tokens));
+        let added_b = count_added(&t.sequence_b, Some(&t.special_tokens));
+        Self {
+            sequence_a: t.sequence_a,
+            sequence_b: t.sequence_b,
+            added_a,
+            added_b,
+            special_tokens: t.special_tokens,
+        }
+    }
+}
+
+/// Count the number of added tokens in the given template
+fn count_added(container: &Template, special_tokens: Option<&Tokens>) -> usize {
+    container
+        .0
+        .iter()
+        .map(|p| match p {
+            Piece::Sequence { .. } => 0,
+            Piece::SpecialToken { id } => {
+                special_tokens.map_or(0, |spt| spt.0.get(id).map_or(0, |s| s.ids.len()))
+            }
+        })
+        .sum()
+}
+
+impl TemplateProcessingBuilder {
+    fn default_seq(&self, type_id: u32) -> Template {
+        Template(vec![Piece::Sequence { type_id }])
+    }
+
+    fn default_added(&self, is_a: bool) -> usize {
+        let container = if is_a {
+            self.sequence_a.as_ref()
+        } else {
+            self.sequence_b.as_ref()
+        };
+        container.map_or(0, |pieces| {
+            count_added(pieces, self.special_tokens.as_ref())
+        })
+    }
+
+    fn validate(&self) -> std::result::Result<(), String> {
+        let check = |sp| {
+            let exist = self
+                .special_tokens
+                .as_ref()
+                .map_or(false, |map| map.0.contains_key(sp));
+            match exist {
+                true => Ok(()),
+                false => Err(format!("The special token `{}` is missing", sp)),
```

Comment:
Could we delay the error to gather all missing tokens ?

As a user I found that having the change tokens one by one was tedious. (Maybe only display the first 10 max to prevent flooding though if case can happen.)